controllers/usernamespace/controller.go

```diff
@@ -0,0 +1,392 @@
+	list := v1.CheClusterList{}
+	if err := r.client.List(ctx, &list, client.InNamespace(namespace)); err != nil {
+		return false
+	}
+
+	return len(list.Items) > 0
+}
+
+func asReconcileRequestForNamespace(obj metav1.Object) []reconcile.Request {
+	return []reconcile.Request{
+		{
+			NamespacedName: types.NamespacedName{Name: obj.GetNamespace()},
+		},
+	}
+}
+
+func asReconcileRequestForNamespaceIf(obj metav1.Object, predicate func() bool) []reconcile.Request {
+	if predicate() {
+		return asReconcileRequestForNamespace(obj)
+	} else {
+		return []reconcile.Request{}
+	}
+}
+
+func (r *CheUserNamespaceReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	ctx := context.Background()
+
+	info, err := r.namespaceCache.ExamineNamespace(ctx, req.Name)
+	if err != nil {
+		logrus.Errorf("Failed to examine namespace %s for presence of Che user info labels: %v", req.Name, err)
+		return ctrl.Result{}, err
+	}
+
+	if info == nil || info.OwnerUid == "" {
+		// we're not handling this namespace
+		return ctrl.Result{}, nil
+	}
+
+	checluster := findManagingCheCluster(*info.CheCluster)
+	if checluster == nil {
+		return ctrl.Result{Requeue: true}, nil
+	}
+
+	if devworkspace.GetDevworkspaceState(r.scheme, checluster) != devworkspace.DevworkspaceStateEnabled {
+		return ctrl.Result{}, nil
+	}
+
+	// let's construct the deployContext to be able to use methods from v1 operator
+	deployContext := &deploy.DeployContext{
+		CheCluster: org.AsV1(checluster),
+		ClusterAPI: deploy.ClusterAPI{
+			Client:          r.client,
+			NonCachedClient: r.client,
+			DiscoveryClient: nil,
+			Scheme:          r.scheme,
+		},
+	}
+
+	if err = r.reconcileSelfSignedCert(ctx, deployContext, req.Name, checluster); err != nil {
+		logrus.Errorf("Failed to reconcile self-signed certificate into namespace '%s': %v", req.Name, err)
+		return ctrl.Result{}, err
+	}
+
+	if err = r.reconcileTrustedCerts(ctx, deployContext, req.Name, checluster); err != nil {
+		logrus.Errorf("Failed to reconcile self-signed certificate into namespace '%s': %v", req.Name, err)
+		return ctrl.Result{}, err
+	}
+
+	if err = r.reconcileProxySettings(ctx, req.Name, checluster, deployContext); err != nil {
+		logrus.Errorf("Failed to reconcile proxy settings into namespace '%s': %v", req.Name, err)
+		return ctrl.Result{}, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func findManagingCheCluster(key types.NamespacedName) *v2alpha1.CheCluster {
+	instances := devworkspace.GetCurrentCheClusterInstances()
+	if len(instances) == 0 {
+		return nil
+	}
+
+	if len(instances) == 1 {
+		for k, v := range instances {
+			if key.Name == "" || (key.Name == k.Name && key.Namespace == k.Namespace) {
+				return &v
+			}
+			return nil
+		}
+	}
+
+	ret, ok := instances[key]
+
+	if ok {
+		return &ret
+	} else {
+		return nil
+	}
+}
+
+func (r *CheUserNamespaceReconciler) reconcileSelfSignedCert(ctx context.Context, deployContext *deploy.DeployContext, targetNs string, checluster *v2alpha1.CheCluster) error {
```

Comment:
I know, these methods (`reconcileTrustedCerts`, `reconcileSelfSignedCert`, ...) are not too big  but in the end it blows controller.go and later it will grow and grow. Which I think makes it more difficult to understand what is supposed to happen here.
So, I think it makes sense to move them to separate three files (trustedCerts, selfSignedCerts, proxySettings), but it's up to you.

Reply:
Agreed, but we can easily move then later, once it becomes a problem.